### ISSUE TITLE
SSE4.1

### ISSUE THREAD

**Issue by mirh** [saviosg/opemu-linux#1]
I'm not sure which kind of issues there would be with 呆丸廖添丁's module
Anyway, did you check these other threads?
https://www.insanelymac.com/forum/topic/322370-task-force-amd-to-missing-sse41-instruction-set-older-cpus-on-macos-sierra-1012x/
https://www.insanelymac.com/forum/topic/337075-new-opemu-supporting-fully-till-sse41/

[OPEMU.zip](https://github.com/saviosg/opemu-linux/files/3988819/OPEMU.zip)

**Comment by mirh** [comment]
I forked your work and tentatively added POPCNT and CRC32. 
If you were caring for SSE4.1, the second thread I posted above should get you similarly handled.

**Comment by Fuseteam** [comment]
@mirh PRs are usually welcome

**Comment by mirh** [comment]
I'd be glad to (for as much my real inexperience)
But if only I knew what OP was trying to do in the first place...

**Comment by Fuseteam** [comment]
the OP? as in this repo's author?

**Comment by mirh** [comment]
Yes. Like, I for one was just interested in those two instructions (popcnt isn't even a simd extension then). But to be sure I'd love to help extending this with customizability and why not even AVX while we are at it. 

While @saviosg seemed more interested to SSSE3 through SSE4.2. But I just have those few lines of readme to imagine his use case, and after 3 months he's still missing.

**Comment by Fuseteam** [comment]
he's basically attempting to use this to get [anbox to work on non-ssse3 hardware](https://github.com/anbox/anbox/issues/499)

**Comment by mirh** [comment]
Oh, lol. Thank you for the context. 
So I suppose this is something @maurossi could also be interested in. 
I'll try to see if I can get some benchmark between his solution, and [opemu](https://web.archive.org/web/20110711112417if_/http://xnu-dev.googlecode.com/files/SSE3%20Emulator.pdf) (and then perhaps I can half-ass together SSE4.1 just the same I already did with the remainder)

**Comment by Fuseteam** [comment]
no problem i just happened to be in the anbox group where he is

**Comment by mirh** [comment]
So... I used [this benchmark](http://www.dalkescientific.com/writings/diary/archive/2011/11/02/faster_popcount_update.html) for popcnt (since it's the only missing instruction on my cpu that works in the emulator too).

Of all the different implementations it tries (possibly because they are performing more than a single call altogether?) I found out only "Wegner/Kernigan" and gcc's builtins were meaningfully impacted. 

"My" OPEMU takes respectively a 60x and 350x performance hit in those cases (compared to the "just software" times - it would be even higher compared to the real speedy hw instruction). While the SSEPlus-based thingy of the "official" [x86](https://groups.google.com/d/msg/android-x86/gfoWKeVUNsA/jiIx1_s89vQJ) [patches](https://github.com/JideTechnology/remixos-kernel/tree/jide_x86_marshmallow_kernel-4.4/arch/x86/kernel) is only 30x on the first. *On the other hand though*, it also simply crashed on the second test. So any kind of "optimization" may as well just be skipping something actually important. 

As a matter of fact intel's sde is just 3x and 22x slower then, but I'm not sure how its userspace runtime patching approach would play out with containers and special cases in general (for example, I reckon vmprotect makes it fail)

...
As for getting SSE4 included instead, it's kinda more difficult than I thought, since AnV decided to reimplement *everything* on top of SSEPlus (especially so if you are just blindly copy-pasting pieces of code here and there)

**Comment by Fuseteam** [comment]
oof this is gonna be diffy
